### ISSUE TITLE
Decompression yields nothing

### ISSUE THREAD

**Issue by NPatch** [gibbed/XCompression#1]
Hi,
I'm trying to decompress an XNB, compiled for XNA 3.1, for a VideoSprite(compressed data should only give a header with info about the accompanying wmv). I've found the XNB source format for 4.0 and so far it matches. I'm feeding the Decompress func, a byte array with just the compressed data(at least wrt the 4.0 format) and the output is zeros.

I only have the XNA redistributable installed for 3.1

Am I missing sth? Does it need the XSDK to do that?

**Comment by gibbed** [comment]
What's your decompression code actually look like?

The specific versions shouldn't matter, compression is same between them.

**Comment by NPatch** [comment]
Code is very lazily written.

```
                uint compressed_file_size = br.ReadUInt32();
                Console.WriteLine("Compressed File Size: {0} bytes", compressed_file_size);
                uint decompressed_data_size = br.ReadUInt32();
                Console.WriteLine("Decompressed Data Size: {0} bytes", decompressed_data_size);

                //Don't know about window or chunk sizes so use defaults
                DecompressionContext dcmp_ctx = new DecompressionContext();

                byte[] compressed_bytes = new byte[compressed_file_size - (uint)ms.Position]; //Calculating compressed data length
                int compressed_bytes_count = compressed_bytes.Length;
                byte[] decompressed_bytes = new byte[decompressed_data_size];
                int decompressed_bytes_count = (int)decompressed_data_size;
                Array.Copy(file_bytes, (int)ms.Position, compressed_bytes, 0, compressed_bytes_count); //copying compressed data into their own buffer

                ErrorCode errc = dcmp_ctx.Decompress(compressed_bytes, 0, ref compressed_bytes_count, decompressed_bytes, 0, ref decompressed_bytes_count);
                if (errc != ErrorCode.None)
                {
                    Console.Error.WriteLine("Decompression failed.");
                }
```

Also tried giving the full xnb byte array at first, using MemoryStream.Position as the offset, but had the same results.

Debugging, I see XnaNative.dll is being found from the redist of 3.1.

**Comment by NPatch** [comment]
Btw the ErrorCode has a value of -2147467259(0x80004005), if that means anything.

**Comment by gibbed** [comment]
If you get an error the output is definitely going to be all zeroes or incomplete decompressed bytes. I want to say the defaults for the window size and chunk size are the XNA defaults, but I'll have to double-check.

**Comment by NPatch** [comment]
Btw I have a question. I'm a total noob wrt to XNA. I've inherited an old project for which I have to update content every now and then and I'm trying to evaluate whether I can automate the full process since it only supports two types of media. The question is, can a Release build of the app work with Debug built XNBs? i.e. avoid any kind of compression issues and only focus on generating bytes for the actual content. Is that possible? I expect not, but asking anyhow.

**Comment by gibbed** [comment]
Can you try these settings and see if it results in successful decompression?

`DecompressionContext dcmp_ctx = new DecompressionContext(0x10000, 0x80000);`

Additionally, `DecompressionContext` is an `IDisposable` and should be wrapped in `using` block or otherwise properly disposed of when done with it.

**Comment by gibbed** [comment]
I actually don't know about that one, this project was built to interact with content that wasn't from XNA originally.

I haven't really touched upon XNA's internal formats myself, my guess is it probably would work, but that's a very uninformed guess.

You may want to consider checking out the following projects:

https://github.com/0x0ade/XnaToFna/ 
https://github.com/FNA-XNA/FNA

**Comment by NPatch** [comment]
YES! They work. I got back Type Reader info!
I'm calling Dispose at the end manually :P.

I know about FNA and XnaToFna but trying to see if I can avoid getting into porting or changing anything at the endsite.

I found this : [https://docplayer.net/49383763-Microsoft-xna-game-studio-4-0-compiled-xnb-content-format.html](https://docplayer.net/49383763-Microsoft-xna-game-studio-4-0-compiled-xnb-content-format.html)

I managed to parse the xnb properly up to the point of the compressed data, using the format above. Now I see I can decompress them. So next step will be trying to compress an uncompressed version and seeing if they match. If so, I only need to see my two use cases and generate the same data for different inputs.

I'll try the uncompressed version too(VS Debug Content built xnbs), but I don't have access to it now, which is why I asked the above.

**Comment by gibbed** [comment]
Great. I'll see about updating XCompression with relevant documentation (comments or otherwise) about the defaults. It turns out the defaults I've given were probably for general use XMemCompress, and that XNA uses slightly different settings.

**Comment by gibbed** [comment]
Also, I didn't mean so much as using those projects directly, but using them for documentation of the underlying formats.

**Comment by NPatch** [comment]
>  I'll see about updating XCompression with relevant documentation (comments or otherwise) about the defaults

Awesome. Plz also comment on that zeros on the output is part of an error during decompression. At first I didn't notice the ErrorCode and thought I'd just get wrong decompressed data back, but I've only gotten back zeroes, no matter how many runs. 

> Also, I didn't mean so much as using **those projects** directly, but using them for documentation of the underlying formats.

Not sure what you mean here.

> It turns out the defaults I've given were probably for general use XMemCompress, and that XNA uses slightly different settings.

Actually the defaults might be an Xbox thing or a WinPhone. I've read that xnbs produced for Xbox don't work on Windows, hence the byte for the target platform. Perhaps the compression window and chunk size changes per target platform.

**Comment by gibbed** [comment]
In the absence of the documentation that you linked. I assume FNA can load existing XNA formats, so its code would be a good reference for the content data formats.

**Comment by NPatch** [comment]
Yes you are right. Btw, there is a C++ XNBParser with most of how the underlying content is structured for each data type(needs an uncompressed XNB tho), which was supposedly given along side the documentation I linked, but it's hard to find(found it on a very random site, most of XNA's documentation has been wiped out, even MS links are hard to google).

**Comment by NPatch** [comment]
Might be a stupid question, but doesn't WindowSize and Chunk Size have to be the same on Compress and Decompress?
I just tried to recreate the xnb and I'm getting more compressed data. Btw if I read the new xnb with the same parser I've written and try to export the texture data into a png, I get a larger PNG(probably because I'm using Bitmap for the conversion and there's no PNG compression), but the image is correct. The headers are all correct. The only thing that's different is the compression size.

**Comment by NPatch** [comment]
Tried 0x8000 as window size and default chunk size for compression and it's 3KB larger.
With 0x10000 and default chunk size, I get 15KB less.

Default Chunk Size is the same as the one you told me for Decompression above, 0x80000.

**Comment by gibbed** [comment]
The window size and chunk size need to match between compression and decompression, yes.

For XNA it needs to be window size=`0x10000`, chunk size=`0x80000`.

As far as compressed data being different in size, assuming the uncompressed bytes are _exactly the same_, it's probably a difference in algorithm changes between builds of XNA.

**Comment by NPatch** [comment]
You mean I have a different build of the 3.1 redist? Interesting.
Ok I'll check it out.

**Comment by gibbed** [comment]
Unlikely, there's only one version of 3.1 that I know of.

https://github.com/gibbed/XCompression/blob/54408d6b6ed39adfb4267f8d0d0879aa5578dd6b/XnaNative.cs#L47-L78

**Comment by gibbed** [comment]
XCompression attempts to load XnaNative from XNA 4.0, then XNA 3.1, then XNA 3.0.

**Comment by NPatch** [comment]
I'll look it up anyhow(probably tomorrow). Not sure how far back the PC that produced the sample xnb was set up. Might be using an older redist from keeping the msi around and reinstalling it(when the owner switched to Win10).

UPDATE: I just can't figure out what else could be the issue. Contrary to before with the Decompress call, Compress has no error, writes on a large buffer, just as much as it needs(the rest are 0) and yields back the correct compressed size(since the one I'm providing it is the fully uncompressed size). Also the headers all seem to decompress correctly, since I then get the same values from parsing the recompressed xnb and if I try to dump a png out of the image data, it's the same image.

**Comment by NPatch** [comment]
First off, I managed to reconstruct an identical version of the Debug Texture XNB, to the byte.
As for the Release version, the differences lie in the compressed part of the XNB as well as the overall file size(including the header). Uncompressed data size of the compressed part is the same in both files. BUT, in the sample XNB, there are 22670bytes that do not make it in the recompressed xnb.
I used Decompress in both xnbs and dumped the decompressed bytes into a file. They are identical. Therefore, somehow Compress is not compressing everything for some reason. Might be a dumb question(I don't know how the compression works), is there a chance Compress might be dropping/discarding bytes if they don't somehow fit into window or chunk sizes? I doubt it, but since I'm using the same configuration as Decompress which seems to work, I can't think of another culprit.

I also got hold of the XnaNative dll from the sample xnb's host pc and it's the same build as the one you have above. On the other hand, that machine also has the Xna Game Studio 3.1 installed so I wonder if there's another version of Compress in there.

I also noticed XnaMediaHelpers_1.dll going through the MSBuild BuildContent targets, though I have no idea if that factors in somehow or if it's irrelevant.

**Comment by NPatch** [comment]
Managed to reproduce the Debug Video XNB. Now looking into the compressed version.
XNBs for WMV files tend to be about ~204bytes all in all. Just a video metadata descriptor which gives video info like wmv filename, duration, resolution etc.
Decompression uses the same configuration as before(window,chunk) and produces correct metadata. But when I try to compress using the same configuration, Compress returns 0 bytes and ErrorCode is None.
